core/base/single_thread_dispatcher.cc

```diff
@@ -0,0 +1,54 @@
+#include "single_thread_dispatcher.h"
+#include <iostream>
+namespace foxglove {
+namespace windows {
+
+SingleThreadDispatcher::SingleThreadDispatcher()
+    : message_window_(
+          std::make_unique<MessageWindow>([this]() { ProcessTasks(); })) {}
+
+SingleThreadDispatcher::~SingleThreadDispatcher() {
+  // TODO: make sure all tasks are done
+}
+
+/**
+ * Call this method from the same thread it was created
+*/
+void SingleThreadDispatcher::Terminate() {
+  if(terminated_.load()) {
+    return;
+  }
+
+  terminated_.store(true);
+ // execute all pending tasks
+  ProcessTasks();
+}
+
+void SingleThreadDispatcher::Dispatch(Task task) {
+  if(terminated_.load()) {
```

Comment:
Can’t we make `terminated_` non-atomic given that we need the mutex anyway?

Reply:
uses now lock + normal bool